[AIRFLOW-4465] ssh_operator: pass remote_host via ssh_hook constructor

eepstein opened the pull request:
The default method of overriding the remote_host for a SSHHook connection  that is creating in the operator should be to do so via the SSHHook constructor rather than updating an instance variable after creation.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

The code path is unchanged and already covered by existing unit tests.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

eepstein committed:
pass remote_host via constructor if defined 

The default method of overriding the remote_host for a SSHHook connection  that is creating in the operator should be to do so via the SSHHook constructor rather than updating an instance variable after creation.

ashb commented on airflow/contrib/operators/ssh_operator.py:
```diff
@@ -71,11 +71,19 @@ def execute(self, context):
                     self.log.info("ssh_conn_id is ignored when ssh_hook is provided.")
+                    
+                if self.remote_host is not None:
+                    self.ssh_hook.remote_host = self.remote_host
```
I think in this case we should ignore the remote_host parameter too.

eepstein replied:
Not so.  This is a feature of ssh_operator that allows re-using an SSHHook (for example, across operators) and dynamically setting the remote_host.  It does have some risk in a multi-threaded execution, so... user beware.  
That said, the indentation was incorrect.  The condition applies IFF ssh_hook is defined on the ssh_operator instance.

ashb commented on airflow/contrib/operators/ssh_operator.py:
```diff
@@ -71,11 +71,19 @@ def execute(self, context):
                                   "Trying ssh_conn_id to create SSHHook.")
-                    self.ssh_hook = SSHHook(ssh_conn_id=self.ssh_conn_id,
-                                            timeout=self.timeout)
+                    if self.remote_host is not None:
```
We don't need this condition - passing None is the same as not passing it at all in this case.

eepstein replied:
good point.

eepstein committed:
Updated per commit comments and flake8. 

Fixes [AIRFLOW-4465](https://issues.apache.org/jira/browse/AIRFLOW-4465).

eepstein committed:
[AIRFLOW-4465] ssh_operator: remote_host in ssh_hook constructor

following existing syntax of checking objects as booleans rather than via "is (not) None".

eepstein committed:
[AIRFLOW-4465] ssh_operator: pass remote_host in

fixing typo and indentation (problem with editing in a browser window)
